libr/core/canal.c

```diff
@@ -2753,39 +2753,106 @@ R_API void r_core_recover_vars(RCore *core, RAnalFunction *fcn, bool argonly) {
 	return;
 }
 
-R_API RList* r_core_anal_graph_to(RCore *core, ut64 addr, int n) {
-	RAnalBlock *bb, *root = NULL, *dest = NULL;
-	RListIter *iter, *iter2;
-	RList *list2 = NULL, *list = NULL;
-	RAnalFunction *fcn;
+static bool anal_path_exists(RCore *core, ut64 from, ut64 to, RList *bbs, int depth, HtUP *state) {
+	RAnalBlock *bb = r_anal_bb_from_offset (core->anal, from);
+	RListIter *iter = NULL;
+	RList *refs = NULL;
+	RAnalRef *refi;
+	RAnalFunction *cur_fcn = NULL;
 
-	r_list_foreach (core->anal->fcns, iter, fcn) {
-		if (!r_anal_fcn_is_in_offset (fcn, core->offset)) {
-			continue;
-		}
-		r_list_foreach (fcn->bbs, iter2, bb) {
-			if (r_anal_bb_is_in_offset (bb, addr)) {
-				dest = bb;
-			}
-			if (r_anal_bb_is_in_offset (bb, core->offset)) {
-				root = bb;
-				r_list_append (list, list2);
+	if (depth < 1) {
+		eprintf ("going too deep\n");
+		return false;
+	}
+
+	r_return_val_if_fail (bbs, false);
+
+	if (!bb) {
+		return false;
+	}
+
+	if (bb->addr != from) {
+		RAnalBlock *_bb = R_NEW0 (RAnalBlock);
+		memcpy (_bb, bb, sizeof (RAnalBlock));
+		_bb->addr = from;
+		bb = _bb;
+	}
+
+	ht_up_update (state, from, bb);
+
+	// try to find the target in the current function
+	if (r_anal_bb_is_in_offset (bb, to) ||
+		((!ht_up_find (state, bb->jump, NULL) && anal_path_exists (core, bb->jump, to, bbs, depth - 1, state))) ||
+		((!ht_up_find (state, bb->fail, NULL) && anal_path_exists (core, bb->fail, to, bbs, depth - 1, state)))) {
+		r_list_prepend (bbs, bb);
+		return true;
+	}
+
+	// find our current function
+	cur_fcn = r_anal_get_fcn_in (core->anal, from, 0);
+
+	// get call refs from current basic block and find a path from them
+	if (cur_fcn) {
+		refs = r_anal_fcn_get_refs (core->anal, cur_fcn);
+		if (refs) {
+			r_list_foreach (refs, iter, refi) {
+				if (refi->type == R_ANAL_REF_TYPE_CALL) {
+					if (r_anal_bb_is_in_offset (bb, refi->at)) {
+						if ((refi->at != refi->addr) && !ht_up_find (state, refi->addr, NULL) && anal_path_exists (core, refi->addr, to, bbs, depth - 1, state)) {
+							r_list_prepend (bbs, bb);
+							return true;
+						}
+					}
+				}
 			}
 		}
 	}
-	if (root && dest) {
-		if (dest == root) {
-			eprintf ("Source and destination are the same\n");
-			return NULL;
+
+	return false;
+}
+
+static RList* anal_graph_to(RCore *core, ut64 addr, int depth) {
+	RAnalFunction *cur_fcn = r_anal_get_fcn_in (core->anal, core->offset, 0);
+	RList *xrefs;
+	RList *list = r_list_new ();
+	HtUP *state = ht_up_new0 ();
+	RListIter *iter = NULL;
+	RAnalRef *xref = NULL;
+
+	if (!list || !state || !cur_fcn) {
+		return NULL;
+	}
+	xrefs = r_anal_xrefs_get (core->anal, cur_fcn->addr);
+
+	// forward search
+	if (anal_path_exists (core, core->offset, addr, list, depth - 1, state)) {
+		ht_up_free (state);
+		return list;
+	}
+
+	// backward search
+	if (xrefs) {
```

Comment:
IMHO this seems like a big hack... Again, as said in one of my first messages, the "right" thing to do here would be to create a graph that connects all the basic blocks and then look for a path that connects two nodes in the graph with DFS.

Reply:
I understand your point, I just needed this feature quickly. and I didn't have enough time to look at the DFS api and understand what is missing for the graph to be able to generate a supergraph of all functions. I can have a look at it, but later.

Reply:
Yep it’s fine as is for now :) , I’ll re-look at the pr on Monday and test it. Thanks!!

Reply:
cool !